File: src/test/java/com/example/the_machine/common/SerializationTest.java
package com.example.the_machine.common;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.example.the_machine.domain.ArtifactEntity;
import com.example.the_machine.dto.AuthorType;
import com.example.the_machine.dto.RunStatus;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.time.Instant;
import java.util.Map;
import org.junit.jupiter.api.Test;

class SerializationTest {

  private final ObjectMapper objectMapper = Json.getObjectMapper();

  @Test
  void enums_serializeAsUpperSnakeCase() throws Exception {
    // Test RunStatus enum
    final var runStatus = RunStatus.REQUIRES_ACTION;
    final var runStatusJson = objectMapper.writeValueAsString(runStatus);
    assertEquals("\"REQUIRES_ACTION\"", runStatusJson);
    System.out.println("[DEBUG_LOG] RunStatus.REQUIRES_ACTION serialized as: " + runStatusJson);

    // Test AuthorType enum
    final var authorType = AuthorType.ASSISTANT;
    final var authorTypeJson = objectMapper.writeValueAsString(authorType);
    assertEquals("\"ASSISTANT\"", authorTypeJson);
    System.out.println("[DEBUG_LOG] AuthorType.ASSISTANT serialized as: " + authorTypeJson);

    // Test ArtifactEntity.Status enum
    final var artifactStatus = ArtifactEntity.Status.PENDING;
    final var artifactStatusJson = objectMapper.writeValueAsString(artifactStatus);
    assertEquals("\"PENDING\"", artifactStatusJson);
    System.out.println(
        "[DEBUG_LOG] ArtifactEntity.Status.PENDING serialized as: " + artifactStatusJson);

    // Test that all enums follow UPPER_SNAKE_CASE pattern
    for (RunStatus status : RunStatus.values()) {
      final var json = objectMapper.writeValueAsString(status);
      final var enumName = status.name();
      assertTrue(enumName.matches("^[A-Z_]+$"),
          "Enum " + enumName + " should be UPPER_SNAKE_CASE");
      assertEquals("\"" + enumName + "\"", json);
    }

    System.out.println("[DEBUG_LOG] All enums serialize as UPPER_SNAKE_CASE strings");
  }

  @Test
  void instant_serializesAsIso8601() throws Exception {
    // Create a test instant
    final var testInstant = Instant.parse("2025-08-20T18:53:00.123456Z");

    // Serialize the instant
    final var instantJson = objectMapper.writeValueAsString(testInstant);
    System.out.println("[DEBUG_LOG] Instant serialized as: " + instantJson);

    // Verify it's in ISO-8601 format (should be quoted string, not timestamp)
    assertFalse(instantJson.matches("^\\d+$"), "Instant should not serialize as timestamp");
    assertTrue(instantJson.startsWith("\"") && instantJson.endsWith("\""),
        "Instant should serialize as quoted string");

    // Remove quotes for format validation
    final var instantString = instantJson.substring(1, instantJson.length() - 1);

    // Verify ISO-8601 format pattern
    assertTrue(instantString.matches("\\d{4}-\\d{2}-\\d{2}T\\d{2}:\\d{2}:\\d{2}(\\.\\d+)?Z"),
        "Instant should be in ISO-8601 format: " + instantString);

    // Test round-trip to ensure deserialization works
    final var deserializedInstant = objectMapper.readValue(instantJson, Instant.class);
    assertEquals(testInstant, deserializedInstant);

    System.out.println(
        "[DEBUG_LOG] Instant round-trip successful: " + testInstant + " -> " + deserializedInstant);
  }

  @Test
  void complexObject_withEnumsAndInstant_serializedCorrectly() throws Exception {
    // Create a test object with enum and instant
    final var now = Instant.parse("2025-08-20T18:53:15.789Z");

    final var testObject = Map.of(
        "status", RunStatus.COMPLETED,
        "author", AuthorType.USER,
        "timestamp", now,
        "message", "Test message"
    );

    // Serialize to JSON
    final var json = objectMapper.writeValueAsString(testObject);
    System.out.println("[DEBUG_LOG] Complex object JSON: " + json);

    // Parse back to verify structure
    final var node = objectMapper.readTree(json);

    // Verify enum serialization (Jackson adds type info as arrays)
    final var statusNode = node.get("status");
    if (statusNode.isArray()) {
      assertEquals("COMPLETED", statusNode.get(1).asText());
    } else {
      assertEquals("COMPLETED", statusNode.asText());
    }

    final var authorNode = node.get("author");
    if (authorNode.isArray()) {
      assertEquals("USER", authorNode.get(1).asText());
    } else {
      assertEquals("USER", authorNode.asText());
    }

    // Verify instant serialization (Jackson adds type info as arrays)
    final var timestampNode = node.get("timestamp");
    final var timestampStr = timestampNode.isArray()
        ? timestampNode.get(1).asText()
        : timestampNode.asText();
    assertTrue(timestampStr.matches("\\d{4}-\\d{2}-\\d{2}T\\d{2}:\\d{2}:\\d{2}(\\.\\d+)?Z"),
        "Timestamp should be ISO-8601 format");

    // Verify other field
    assertEquals("Test message", node.get("message").asText());

    System.out.println("[DEBUG_LOG] Complex object serialization verified");
  }
}

File: src/test/kotlin/com/example/the_machine/common/SerializationTest.kt
package com.example.the_machine.common

import com.example.the_machine.config.TestDataConfiguration
import com.example.the_machine.domain.ArtifactEntity
import com.example.the_machine.dto.AuthorType
import com.example.the_machine.dto.RunStatus
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.encodeToJsonElement
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import java.time.Instant

class SerializationTest {

  private val json = KotlinSerializationConfig.staticJson

  @Test
  fun enums_serializeAsUpperSnakeCase() {
    // Test RunStatus enum
    val runStatus = RunStatus.REQUIRES_ACTION
    val runStatusJson = json.encodeToString<RunStatus>(runStatus)
    assertEquals("\"REQUIRES_ACTION\"", runStatusJson)
    println("[DEBUG_LOG] RunStatus.REQUIRES_ACTION serialized as: $runStatusJson")

    // Test AuthorType enum
    val authorType = AuthorType.ASSISTANT
    val authorTypeJson = json.encodeToString<AuthorType>(authorType)
    assertEquals("\"ASSISTANT\"", authorTypeJson)
    println("[DEBUG_LOG] AuthorType.ASSISTANT serialized as: $authorTypeJson")

    // Test ArtifactEntity.Status enum
    val artifactStatus = ArtifactEntity.Status.PENDING
    val artifactStatusJson = json.encodeToString<ArtifactEntity.Status>(artifactStatus)
    assertEquals("\"PENDING\"", artifactStatusJson)
    println("[DEBUG_LOG] ArtifactEntity.Status.PENDING serialized as: $artifactStatusJson")

    // Test that all enums follow UPPER_SNAKE_CASE pattern
    for (status in RunStatus.entries) {
      val statusJson = json.encodeToString<RunStatus>(status)
      val enumName = status.name
      assertTrue(
        enumName.matches(Regex("^[A-Z_]+$")),
        "Enum $enumName should be UPPER_SNAKE_CASE"
      )
      assertEquals("\"$enumName\"", statusJson)
    }

    println("[DEBUG_LOG] All enums serialize as UPPER_SNAKE_CASE strings")
  }

  @Test
  fun instant_serializesAsIso8601() {
    // Create a test instant
    val testInstant = TestDataConfiguration.FIXED_INSTANT

    // Serialize the instant
    val instantJson = json.encodeToString(testInstant)
    println("[DEBUG_LOG] Instant serialized as: $instantJson")

    // Verify it's in ISO-8601 format (should be quoted string, not timestamp)
    assertFalse(instantJson.matches(Regex("^\\d+$")), "Instant should not serialize as timestamp")
    assertTrue(
      instantJson.startsWith("\"") && instantJson.endsWith("\""),
      "Instant should serialize as quoted string"
    )

    // Remove quotes for format validation
    val instantString = instantJson.substring(1, instantJson.length - 1)

    // Verify ISO-8601 format pattern
    assertTrue(
      instantString.matches(Regex("\\d{4}-\\d{2}-\\d{2}T\\d{2}:\\d{2}:\\d{2}(\\.\\d+)?Z")),
      "Instant should be in ISO-8601 format: $instantString"
    )

    // Test round-trip to ensure deserialization works
    val deserializedInstant = json.decodeFromString<Instant>(instantJson)
    assertEquals(testInstant, deserializedInstant)

    println("[DEBUG_LOG] Instant round-trip successful: $testInstant -> $deserializedInstant")
  }

  @Test
  fun complexObject_withEnumsAndInstant_serializedCorrectly() {
    // Create a test object with enum and instant
    val now = TestDataConfiguration.FIXED_INSTANT_LATER

    val testObject = buildJsonObject {
      put("status", json.encodeToJsonElement(RunStatus.COMPLETED))
      put("author", json.encodeToJsonElement(AuthorType.USER))
      put("timestamp", json.encodeToJsonElement(now))
      put("message", json.encodeToJsonElement("Test message"))
    }

    // Serialize to JSON
    val jsonString = json.encodeToString(testObject)
    println("[DEBUG_LOG] Complex object JSON: $jsonString")

    // Parse back to verify structure
    val jsonElement = json.parseToJsonElement(jsonString)
    val jsonObj = jsonElement.jsonObject

    // Verify enum serialization
    val statusValue = jsonObj["status"]?.jsonPrimitive?.content
    assertEquals("COMPLETED", statusValue)

    val authorValue = jsonObj["author"]?.jsonPrimitive?.content
    assertEquals("USER", authorValue)

    // Verify instant serialization
    val timestampValue = jsonObj["timestamp"]?.jsonPrimitive?.content
    assertTrue(
      timestampValue?.matches(Regex("\\d{4}-\\d{2}-\\d{2}T\\d{2}:\\d{2}:\\d{2}(\\.\\d+)?Z")) == true,
      "Timestamp should be ISO-8601 format"
    )

    // Verify other field
    assertEquals("Test message", jsonObj["message"]?.jsonPrimitive?.content)

    println("[DEBUG_LOG] Complex object serialization verified")
  }
}